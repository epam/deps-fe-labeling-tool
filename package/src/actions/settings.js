import { createAction } from 'redux-actions'

const FEATURE_NAME = 'SETTINGS'

const storeSettings = createAction(`${FEATURE_NAME}/STORE`)

const resetDefault = createAction(`${FEATURE_NAME}/RESET_DEFAULT`)

export {
  resetDefault,
  storeSettings
}
