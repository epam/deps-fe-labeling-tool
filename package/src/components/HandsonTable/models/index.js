
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
