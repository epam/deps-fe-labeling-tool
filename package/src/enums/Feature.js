const Feature = {
  AUTO_SAVE: 'autoSave',
  EXPORT: 'export',
  IMPORT: 'import',
  PAGING: 'paging',
  LANGUAGE: 'language',
  ROTATION: 'rotation',
  SPECIAL_SYMBOLS: 'specialSymbols',
  MANAGE_FIELDS: 'manageFields',
  MULTI_ASSIGN_LABELS: 'multiAssignLabels',
  MULTI_ASSIGN_TABLES: 'multiAssignTables'
}

export {
  Feature
}
