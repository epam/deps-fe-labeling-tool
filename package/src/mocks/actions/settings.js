const storeSettings = jest.fn((settings) => ({
  type: 'MOCK_STORE_SETTINGS_ACTION',
  payload: settings
}))

const mockSettingsActions = {
  storeSettings
}

export {
  mockSettingsActions
}
