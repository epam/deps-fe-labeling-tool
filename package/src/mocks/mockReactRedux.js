const mockDispatch = jest.fn((action) => action)

const createMockMapStateToProps = (mapStateToProps) => {
  if (!mapStateToProps) {
    return mapStateToProps
  }

  return (state, ownProps) => ({
    props: mapStateToProps(state, ownProps)
  })
}

const getDispatchFunctionFormShorthandObject = (actionsObj) => (dispatch) => (
  Object.keys(actionsObj).reduce(
    (props, actionKey) => ({
      ...props,
      [actionKey]: jest.fn(
        (...args) => dispatch(actionsObj[actionKey].apply(undefined, args))
      )
    }),
    {}
  )
)

const createMockMapDispatchToProps = (mapDispatchToProps) => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps
  }

  const mappingFunction = typeof mapDispatchToProps === 'function'
    ? mapDispatchToProps
    : getDispatchFunctionFormShorthandObject(mapDispatchToProps)

  return (dispatch, ownProps) => {
    dispatch = dispatch || mockDispatch

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    }
  }
}

const createMockMergeProps = (mergeProps) => {
  if (!mergeProps) {
    return mergeProps
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: mockDispatch }

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    }
  }
}

const mockReactRedux = {
  connect: (mapStateToProps, mapDispatchToProps, mergeProps) => (WrappedComponent) => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    WrappedComponent
  }),
  batch: (fn) => Promise.resolve(fn),
  useDispatch: jest.fn(() => mockDispatch),
  useSelector: jest.fn((selector) => selector())
}

export {
  mockReactRedux
}
