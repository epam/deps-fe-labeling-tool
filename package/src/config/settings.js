import { Settings } from '@/models/Settings'

const settings = new Settings()

export {
  settings
}
