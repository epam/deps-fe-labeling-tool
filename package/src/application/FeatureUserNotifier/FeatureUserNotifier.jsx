import { useEffect, useCallback } from 'react'
import { useSelector } from 'react-redux'
import { Feature } from '@/enums/Feature'
import { DEFAULT_AUTO_SAVE_INTERVAL_MS } from '@/models/Settings'
import { settingsSelector } from '@/selectors/settings'
import { getApi } from '@/services/api'
import { childrenShape } from '@/utils/propTypes'
import { replaceAll } from '@/utils/string'

const RESOURCE_AUTO_SAVE_NOTIFICATION_1 =
   'Please be aware, autosaving occurs every {0} minutes'

const FeatureUserNotifier = ({ children }) => {
  const settings = useSelector(settingsSelector)

  const autoSaveNotify = (autoSaveFeature) => {
    const interval =
       autoSaveFeature?.data?.interval ?? DEFAULT_AUTO_SAVE_INTERVAL_MS
    const minutes = interval / 1000 / 60
    getApi().notify?.success?.(
      replaceAll(RESOURCE_AUTO_SAVE_NOTIFICATION_1, minutes)
    )
  }

  const notifyUser = useCallback(() => {
    const autoSaveFeature = settings.features?.find(
      (f) => f.code === Feature.AUTO_SAVE
    )
    autoSaveFeature && autoSaveNotify(autoSaveFeature)
  }, [settings.features])

  useEffect(() => {
    notifyUser()
  }, [notifyUser, settings])

  return children
}

FeatureUserNotifier.propTypes = {
  children: childrenShape.isRequired
}

export { FeatureUserNotifier }
