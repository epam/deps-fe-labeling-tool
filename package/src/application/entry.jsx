import 'antd/lib/style/themes/default.less'
import 'antd/lib/style/core/index.less'
import React from 'react'
import ReactDOM from 'react-dom'
import { LabelingTool } from '@/application/LabelingTool'
import { SHIM_CONFIG } from '@/config/shim'

const LABELING_TOOL_EXPORT_KEY = 'LabelingTool'

const mount = (targetId, config) => {
  ReactDOM.render(
    <LabelingTool config={config} />,
    document.getElementById(targetId)
  )
}

const unmount = (targetId) => ReactDOM.unmountComponentAtNode(
  document.getElementById(targetId)
)

if (window.LABELING_TOOL_SHIM_TARGET_ID) {
  mount(window.LABELING_TOOL_SHIM_TARGET_ID, SHIM_CONFIG)
} else {
  window[LABELING_TOOL_EXPORT_KEY] = {
    mount,
    unmount
  }
}
