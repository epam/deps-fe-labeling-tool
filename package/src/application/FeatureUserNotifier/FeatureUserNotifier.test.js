import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockSettingsSelectors } from '@/mocks/selectors/settings'
import React from 'react'
import { shallow } from 'enzyme'
import { FeatureUserNotifier } from '@/application/FeatureUserNotifier'
import { Feature } from '@/enums/Feature'
import { DEFAULT_AUTO_SAVE_INTERVAL_MS } from '@/models/Settings'
import { settingsSelector } from '@/selectors/settings'

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn((f) => f())
}))
jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/settings', () => mockSettingsSelectors)

const mockNotifySuccess = jest.fn()

jest.mock('@/services/api', () => ({
  getApi: jest.fn(() => ({
    notify: {
      success: mockNotifySuccess
    }
  }))
}))

describe('Application: FeatureUserNotifier', () => {
  const mockChildrenComponent = 'Test children'
  const getMessage = (minutes) => `Please be aware, autosaving occurs every ${minutes} minutes`

  it('should return correct snapshot', () => {
    const wrapper = shallow(<FeatureUserNotifier>{mockChildrenComponent}</FeatureUserNotifier>)
    expect(wrapper).toMatchSnapshot()
  })

  it('should not call notify success if autoSaveFeature is not enabled', () => {
    settingsSelector.mockReturnValueOnce({
      features: []
    })
    shallow(<FeatureUserNotifier>{mockChildrenComponent}</FeatureUserNotifier>)
    expect(mockNotifySuccess).not.toBeCalled()
  })

  const testCases = [
    {
      name:
        'should call notify success first time with the appropriate message and default interval if autoSaveFeature is enabled',
      features: [
        {
          code: Feature.AUTO_SAVE
        }
      ],
      result: getMessage(DEFAULT_AUTO_SAVE_INTERVAL_MS / 1000 / 60)
    },
    {
      name:
        'should call notify success first time with the appropriate message and provided interval if autoSaveFeature is enabled',
      features: [
        {
          code: Feature.AUTO_SAVE,
          data: {
            interval: 60000
          }
        }
      ],
      result: getMessage(1)
    }
  ]

  it.each(testCases)('$name', ({ features, result }) => {
    settingsSelector.mockReturnValueOnce({
      features
    })
    shallow(<FeatureUserNotifier>{mockChildrenComponent}</FeatureUserNotifier>)
    expect(mockNotifySuccess).nthCalledWith(1, result)
  })
})
