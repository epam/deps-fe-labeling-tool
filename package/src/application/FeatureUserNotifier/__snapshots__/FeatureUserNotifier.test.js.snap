// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application: FeatureUserNotifier should return correct snapshot 1`] = `"Test children"`;
